### Môn Lập trình cho khoa học dữ liệu
#### Giảng viên: Trần Trung Kiên


<h1><center>Đồ án cuối kỳ</center></h1>

MSSV |	Họ và tên | Tên tài khoản github
--- | --- | ---
19120186 |	Đỗ Lê Khánh Đăng | dolekhanhdang
19120412 |	Nguyễn Minh Tú   |  tuak123ax

### Về dữ liệu
Dữ liệu này về lương của những công việc liên quan đến data: data scientist, data engineer, .... <br />
Nguồn: https://www.kaggle.com/nikhilbhathi/data-scientist-salary-us-glassdoor. <br />
Trong mục mô tả của dữ liệu:<br />
- The dataset is scraped from the Glassdoor website using Selenium scrapper. After scrapping, the raw dataset was cleaned and made usable for performing data analysis and modelling. The dataset contains information about the minimum salary, maximum salary, average salary, job description, age of the company in years, etc.
- Dữ liệu được lấy bằng cách scraped từ website của Glassdoor với Selenium scrapper. Sau khi lấy, dữ liệu đã được xử lý phù hợp cho phân tích dữ liệu và mô hình hóa. Tập dữ liệu chứa thông tin về mức lương tối thiểu, tối đa, trung bình, mô tả công việc, tuổi đời công ty,..

## Import thư viện

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Khám phá dữ liệu 

### Đọc dữ liệu từ file csv vào dataframe

In [3]:
salary_df = pd.read_csv('data_cleaned_2021.csv',index_col='index')
salary_df.head()

Job Title              Salary Estimate  \
index                                                           
0                 Data Scientist   $53K-$91K (Glassdoor est.)   
1      Healthcare Data Scientist  $63K-$112K (Glassdoor est.)   
2                 Data Scientist   $80K-$90K (Glassdoor est.)   
3                 Data Scientist   $56K-$97K (Glassdoor est.)   
4                 Data Scientist  $86K-$143K (Glassdoor est.)   

                                         Job Description  Rating  \
index                                                              
0      Data Scientist\nLocation: Albuquerque, NM\nEdu...     3.8   
1      What You Will Do:\n\nI. General Summary\n\nThe...     3.4   
2      KnowBe4, Inc. is a high growth information sec...     4.8   
3      *Organization and Job ID**\nJob ID: 310709\n\n...     3.8   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

                                     Company Name         Location  \
index                                                                
0                          Tecolote Research\n3.8  Albuquerque, NM   
1      University of Maryland Medical System\n3.4    Linthicum, MD   
2                                    KnowBe4\n4.8   Clearwater, FL   
3                                       PNNL\n3.8     Richland, WA   
4                         Affinity Solutions\n2.9     New York, NY   

         Headquarters          Size  Founded   Type of ownership  ... tensor  \
index                                                             ...          
0          Goleta, CA   501 - 1000      1973   Company - Private  ...      0   
1       Baltimore, MD       10000+      1984  Other Organization  ...      0   
2      Clearwater, FL   501 - 1000      2010   Company - Private  ...      0   
3        Richland, WA  1001 - 5000      1965          Government  ...      0   
4        New York, NY     51 - 200      1998   Company - Private  ...      0   

      hadoop tableau bi  flink  mongo  google_an   job_title_sim  \
index                                                              
0          0       1  1      0      0          0  data scientist   
1          0       0  0      0      0          0  data scientist   
2          0       0  0      0      0          0  data scientist   
3          0       0  0      0      0          0  data scientist   
4          0       0  0      0      0          0  data scientist   

       seniority_by_title Degree  
index                             
0                      na      M  
1                      na      M  
2                      na      M  
3                      na     na  
4                      na     na  

[5 rows x 41 columns]

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

Tính số dòng và số cột sau đó lưu vào 2 biến num_rows và num_cols.

In [4]:
num_rows=salary_df.shape[0]
num_cols=salary_df.shape[1]
print(num_rows)
print(num_cols)

742
41


Dữ liệu có 742 dòng và 41 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng cho ta biết các thông tin liên quan đến công việc và mức lương của công việc đó. 

Các dòng có ý nghĩa giống nhau.

### Dữ liệu có các dòng bị lặp không?

kiểm tra và lưu kết quả vào biến 'dup'. Biến này sẽ có giá trị là số lượng dòng bị lặp.

In [5]:
dup=salary_df.duplicated().sum()
dup

275

Người đăng tải dữ liệu nói rằng họ đã loại bỏ các dòng trùng nhưng không hiểu sao chúng ta tính vẫn còn các dòng trùng.

Vậy thì chúng ta phải loại bỏ các dòng trùng.

In [6]:
salary_df=salary_df.drop_duplicates()

In [7]:
salary_df.shape

(467, 41)

Lúc này dữ liệu giảm từ 742 dòng xuống còn 467 dòng.

### Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả của về các cột trong file "data_cleaned_2021.csv":

index: index

Job Title: Chức danh của công việc, eg. Data scientist, junior data scientist, senior data scientist etc

Salary Estimate: Mức lương và nguồn

Job Description: Cho biết những phẩm chất mà công ty mong muốn và những gì được mong đợi từ chức danh công việc

Rating: đánh giá của công ty

Company Name: tên công ty

Location: Vị trí của công việc

Headquarters: Địa điểm đặt trụ sở chính của công ty

Size: số lượng nhân viên làm việc trong công ty

Founded: Công ty được thành lập vào năm

Type of ownership: Cho chúng ta biết công ty thuộc sở hữu tư nhân, công cộng hay chính phủ

Industry: Các ngành như công nghiệp CNTT, công nghiệp dược phẩm, v.v.

Sector: Lĩnh vực mà công ty hoạt động

Revenue: Tổng doanh thu của công ty mỗi năm

Competitors: Các đối thủ cạnh tranh hiện tại của công ty

Hourly: 1: Nếu lương được báo cáo theo giờ. 0: Nếu không

Employer provided: 1: Nếu mức lương được cung cấp bởi người sử dụng lao động 0: Nếu không

Lower Salary: Mức lương tối thiểu được báo cáo cho công việc trong một công ty cụ thể

Upper Salary: Mức lương tối đa được báo cáo cho công việc trong một công ty cụ thể

Avg Salary(K): Mức lương trung bình được báo cáo cho công việc trong một công ty cụ thể

company_txt: tên công ty

Job Location: vị trí của công việc (bang nào)

Age: tuổi công ty

Python,spark,aws,excel,sql,sas,keras,pytorch,scikit,tensor,hadoop,tableau,bi,flink,mongo,google_an: các kỹ năng yêu cầu (1: nếu kỹ năng cần thiết 0: nếu ngược lại)

job_title_sim: Chức danh đơn giản

seniority_by_title: Thâm niên trong chức danh

Degree: trình độ (M: Nếu chức danh công việc yêu cầu hoặc cung cấp số năm kinh nghiệm để có được nó, P: Nếu chức danh công việc yêu cầu hoặc cung cấp số năm kinh nghiệm để có được nó)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [8]:
col_dtypes=salary_df.dtypes
col_dtypes

Job Title              object
Salary Estimate        object
Job Description        object
Rating                float64
Company Name           object
Location               object
Headquarters           object
Size                   object
Founded                 int64
Type of ownership      object
Industry               object
Sector                 object
Revenue                object
Competitors            object
Hourly                  int64
Employer provided       int64
Lower Salary            int64
Upper Salary            int64
Avg Salary(K)         float64
company_txt            object
Job Location           object
Age                     int64
Python                  int64
spark                   int64
aws                     int64
excel                   int64
sql                     int64
sas                     int64
keras                   int64
pytorch                 int64
scikit                  int64
tensor                  int64
hadoop                  int64
tableau   

### Cột có dtype là object nghĩa là sao?

Trong Pandas, kiểu dữ liệu object thường ám chỉ chuỗi, nhưng thật ra kiểu dữ liệu object có thể chứa một đối tượng bất kỳ trong Python (vì thật ra ở bên dưới kiểu dữ liệu object chứa địa chỉ). Nếu một cột trong dataframe có dtype là object thì có thể các phần tử trong cột này sẽ có kiểu dữ liệu khác nhau; để biết được kiểu dữ liệu thật sự của các phần tử trong cột này thì ta phải truy xuất vào từng phần tử và dùng câu lệnh type để xem kiểu dữ liệu.

In [9]:
def open_object_dtype(s):
    dtypes = set()

    # YOUR CODE HERE
    s=s.apply(type)
    dtypes.update(s.unique().tolist())
    
    return dtypes

In [10]:
print(open_object_dtype(salary_df['Job Title']))
print(open_object_dtype(salary_df['Salary Estimate']))
print(open_object_dtype(salary_df['Job Description']))
print(open_object_dtype(salary_df['Company Name']))
print(open_object_dtype(salary_df['Location']))
print(open_object_dtype(salary_df['Headquarters']))
print(open_object_dtype(salary_df['Size']))
print(open_object_dtype(salary_df['Type of ownership']))
print(open_object_dtype(salary_df['Industry']))
print(open_object_dtype(salary_df['Sector']))
print(open_object_dtype(salary_df['Revenue']))
print(open_object_dtype(salary_df['Competitors']))
print(open_object_dtype(salary_df['company_txt']))
print(open_object_dtype(salary_df['Job Location']))
print(open_object_dtype(salary_df['job_title_sim']))
print(open_object_dtype(salary_df['seniority_by_title']))
print(open_object_dtype(salary_df['Degree']))

{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}


Vậy là các cột đều có kiểu dữ liệu là chuỗi, chúng ta không cần xử lý gì thêm.

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

Các cột dữ liệu dạng numeric là: Rating, Founded, Lower Salary, Upper Salary, Avg Salary(K), Age

Với mỗi cột numeric ta sẽ tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Sau đó lưu kết quả vào dataframe 'nume_col_df'. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 6 cột là "Rating", "Founded", "Lower Salary", "Upper Salary","Avg Salary(K)","Age".

In [31]:
def XuLy(df,nume_col_df,columnName):
    list_values=[((df[columnName].isna().sum()/df[columnName].size)*100),
    (df[columnName].min()),
    (df[columnName].max())]
    nume_col_df[columnName]=list_values

In [32]:
cols=['Rating', 'Founded', 'Lower Salary', 'Upper Salary','Avg Salary(K)','Age']
rows=['missing_ratio', 'min', 'max']
nume_col_df=pd.DataFrame(index=rows,columns=cols)
XuLy(salary_df,nume_col_df,'Rating')
XuLy(salary_df,nume_col_df,'Founded')
XuLy(salary_df,nume_col_df,'Lower Salary')
XuLy(salary_df,nume_col_df,'Upper Salary')
XuLy(salary_df,nume_col_df,'Avg Salary(K)')
XuLy(salary_df,nume_col_df,'Age')

In [33]:
nume_col_df

Rating  Founded  Lower Salary  Upper Salary  Avg Salary(K)  \
missing_ratio     0.0      0.0           0.0           0.0            0.0   
min              -1.0     -1.0          15.0          16.0           15.5   
max               5.0   2019.0         202.0         306.0          254.0   

                 Age  
missing_ratio    0.0  
min             -1.0  
max            277.0

min của 3 cột Rating, Founded và Age có vẻ bất thường. Giá trị của 3 cột này không thể là số âm được.

Có vẻ -1 tượng trưng cho missing values trong tập dữ liệu này.

### Vấn đề cần tiền xử lý

-Ở 3 cột 'Rating','Founded','Age' chúng ta sẽ loại bỏ những dòng có giá trị -1 vì những dòng này có vẻ tượng trưng cho missing values.

### Tiền xử lý

Ta sẽ loại bỏ các giá trị -1 ở cột 'Rating', 'Founded','Age'.

In [34]:
new_df=salary_df[salary_df['Rating']>-1]
new_df=new_df[new_df['Founded']>-1]
new_df=new_df[new_df['Age']>-1]

Lúc này min của các cột sẽ trở nên bình thường.

In [35]:
cols=['Rating', 'Founded', 'Lower Salary', 'Upper Salary','Avg Salary(K)','Age']
rows=['missing_ratio', 'min', 'max']
nume_col_df=pd.DataFrame(index=rows,columns=cols)
XuLy(new_df,nume_col_df,'Rating')
XuLy(new_df,nume_col_df,'Founded')
XuLy(new_df,nume_col_df,'Lower Salary')
XuLy(new_df,nume_col_df,'Upper Salary')
XuLy(new_df,nume_col_df,'Avg Salary(K)')
XuLy(new_df,nume_col_df,'Age')

In [36]:
nume_col_df

Rating  Founded  Lower Salary  Upper Salary  Avg Salary(K)  \
missing_ratio     0.0      0.0           0.0           0.0            0.0   
min               1.9   1744.0          15.0          16.0           15.5   
max               5.0   2019.0         202.0         306.0          254.0   

                 Age  
missing_ratio    0.0  
min              2.0  
max            277.0

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

Các cột có kiểu dữ liệu dạng categorical là các cột còn lại khác các cột numeric.

Với mỗi cột categorical ta sẽ tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Sau đó lưu kết quả vào dataframe 'cate_col_df'. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 35 cột là các cột khác các cột numeric.

In [37]:
cate_df=new_df.drop(columns=cols)

In [39]:
cate_col_df=pd.DataFrame(index=rows,columns=cate_df.columns)
for i in cate_df.columns:
    XuLy(salary_df,cate_col_df,i)

In [40]:
cate_col_df

Job Title                     Salary Estimate  \
missing_ratio                0.0                                 0.0   
min            Ag Data Scientist    $10-$17 Per Hour(Glassdoor est.)   
max             Web Data Analyst  Employer Provided Salary:$85K-$90K   

                                                 Job Description  \
missing_ratio                                                0.0   
min            (We are unable to sponsor for this role now or...   
max            ­­Missile Systems provides many different high...   

                               Company Name          Location  \
missing_ratio                           0.0               0.0   
min            1-800-FLOWERS.COM, Inc.\n2.8  Agoura Hills, CA   
max                          webfx.com\n4.7     Worcester, MA   

                      Headquarters     Size               Type of ownership  \
missing_ratio                  0.0      0.0                             0.0   
min                             -1  1 - 50             College / University   
max            Zurich, Switzerland  unknown  Subsidiary or Business Segment   

                Industry            Sector  ... tensor hadoop  tableau   bi  \
missing_ratio        0.0               0.0  ...    0.0    0.0      0.0  0.0   
min                   -1                -1  ...    0.0    0.0      0.0  0.0   
max            Wholesale  Travel & Tourism  ...    1.0    1.0      1.0  1.0   

              flink mongo  google_an                   job_title_sim  \
missing_ratio   0.0   0.0        0.0                             0.0   
min             0.0   0.0        0.0  Data scientist project manager   
max             1.0   1.0        1.0                 other scientist   

               seniority_by_title  Degree  
missing_ratio                 0.0     0.0  
min                            jr       M  
max                            sr      na  

[3 rows x 35 columns]

Ta thấy có những cột có giá trị min là -1, điều đó có nghĩa là có missing values trong dữ liệu.

### Vấn đề cần xử lý

-Loại bỏ các dòng có giá trị là -1 ở mỗi cột.

### Tiền xử lý

Ta sẽ loại bỏ tất cả các dòng có giá trị -1 ứng với mỗi cột.

In [41]:
def removeMissing(df,cols_list):
    result_df=df
    for col in cols_list:
        result_df=result_df[result_df[col]!="-1"]
    return result_df

In [42]:
cate_df=removeMissing(cate_df,cate_df.columns)

Lúc này min của các cột sẽ bình thường.

In [44]:
cate_col_df=pd.DataFrame(index=rows,columns=cate_df.columns)
for i in cate_df.columns:
    XuLy(cate_df,cate_col_df,i)

In [49]:
cate_col_df

Job Title  \
missing_ratio                                                0.0   
min            Assistant Director/Director, Office of Data Sc...   
max                                             VP, Data Science   

                            Salary Estimate  \
missing_ratio                           0.0   
min            $100K-$160K (Glassdoor est.)   
max             $98K-$182K (Glassdoor est.)   

                                                 Job Description  \
missing_ratio                                                0.0   
min            (We are unable to sponsor for this role now or...   
max            isn’t your usual company. Our work is powered ...   

                Company Name          Location         Headquarters  \
missing_ratio            0.0               0.0                  0.0   
min            1904labs\n4.7  Agoura Hills, CA          Alameda, CA   
max            comScore\n2.5    Woodbridge, NJ  Zurich, Switzerland   

                    Size               Type of ownership  \
missing_ratio        0.0                             0.0   
min              1 - 50                Company - Private   
max            51 - 200   Subsidiary or Business Segment   

                              Industry               Sector  ... tensor  \
missing_ratio                      0.0                  0.0  ...    0.0   
min            Advertising & Marketing  Aerospace & Defense  ...    0.0   
max                        Video Games     Travel & Tourism  ...    1.0   

              hadoop  tableau   bi flink mongo  google_an    job_title_sim  \
missing_ratio    0.0      0.0  0.0   0.0   0.0        0.0              0.0   
min              0.0      0.0  0.0   0.0   0.0        0.0          analyst   
max              1.0      1.0  1.0   1.0   1.0        1.0  other scientist   

               seniority_by_title  Degree  
missing_ratio                 0.0     0.0  
min                            jr       M  
max                            sr      na  

[3 rows x 35 columns]

#### Vậy là chúng ta đã khám phá và hiểu dữ liệu cũng như tiền xử lý dữ liệu để có thể trả lời cho các câu hỏi.

### Các câu hỏi ý nghĩa cần trả lời